pir/pir-internal/src/main/java/com/duckduckgo/pir/internal/common/PirRunStateHandler.kt

```diff
@@ -0,0 +1,279 @@
+import com.duckduckgo.pir.internal.store.db.PirBrokerScanLog
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.adapters.PolymorphicJsonAdapterFactory
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import javax.inject.Inject
+import kotlinx.coroutines.withContext
+
+interface PirRunStateHandler {
+    suspend fun handleState(pirRunState: PirRunState)
+
+    sealed class PirRunState(open val brokerName: String) {
+        data class BrokerManualScanStarted(
+            override val brokerName: String,
+            val eventTimeInMillis: Long,
+        ) : PirRunState(brokerName)
+
+        data class BrokerManualScanCompleted(
+            override val brokerName: String,
+            val startTimeInMillis: Long,
+            val eventTimeInMillis: Long,
+            val totalTimeMillis: Long,
+            val isSuccess: Boolean,
+        ) : PirRunState(brokerName)
+
+        data class BrokerScheduledScanStarted(
+            override val brokerName: String,
+            val eventTimeInMillis: Long,
+        ) : PirRunState(brokerName)
+
+        data class BrokerScheduledScanCompleted(
+            override val brokerName: String,
+            val startTimeInMillis: Long,
+            val eventTimeInMillis: Long,
+            val totalTimeMillis: Long,
+            val isSuccess: Boolean,
+        ) : PirRunState(brokerName)
+
+        data class BrokerScanActionSucceeded(
+            override val brokerName: String,
+            val pirSuccessResponse: PirSuccessResponse,
+        ) : PirRunState(brokerName)
+
+        data class BrokerScanActionFailed(
+            override val brokerName: String,
+            val actionType: String,
+            val pirErrorReponse: PirErrorReponse,
+        ) : PirRunState(brokerName)
+
+        data class BrokerOptOutStarted(
+            override val brokerName: String,
+        ) : PirRunState(brokerName)
+
+        data class BrokerOptOutCompleted(
+            override val brokerName: String,
+            val extractedProfile: ExtractedProfile,
+            val startTimeInMillis: Long,
+            val endTimeInMillis: Long,
+        ) : PirRunState(brokerName)
+
+        data class BrokerOptOutActionSucceeded(
+            override val brokerName: String,
+            val extractedProfile: ExtractedProfile,
+            val completionTimeInMillis: Long,
+            val actionType: String,
+            val result: PirSuccessResponse,
+        ) : PirRunState(brokerName)
+
+        data class BrokerOptOutActionFailed(
+            override val brokerName: String,
+            val extractedProfile: ExtractedProfile,
+            val completionTimeInMillis: Long,
+            val actionType: String,
+            val result: PirErrorReponse,
+        ) : PirRunState(brokerName)
+    }
+}
+
+@ContributesBinding(AppScope::class)
+class RealPirRunStateHandler @Inject constructor(
```

Comment:
Extracted this class from the runner since that class is getting too big. This handles pixels and saving states to db